File: easy/find-minimum/python/find_minimum.py
from typing import List


def find_minimum(arr: List[int]) -> int:
    """
    Returns the smallest element in the array.

    Args:
        arr (List[int]): List of integers

    Returns:
        int: Minimum value in the array

    Raises:
        ValueError: If the input list is empty
    """
    if not arr:
        raise ValueError("Input array is empty")
    minimum = arr[0]
    for num in arr:
        if num < minimum:
            minimum = num
    return minimum


# Simple test harness
if __name__ == "__main__":
    test_cases = [
        ([3, 1, 4, 1, 5], 1),
        ([10, 20, 5, 30, 2], 2),
        ([7], 7),
        ([5, 5, 5], 5),
        ([-10, 0, 5, -3], -10),
    ]

    for arr, expected in test_cases:
        result = find_minimum(arr)
        assert (
            result == expected
        ), f"Failed for {arr}: expected {expected}, got {result}"
    print("All test cases passed.")

File: easy/find-minimum/cpp/find_minimum.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <stdexcept>
using namespace std;

int findMinimum(const vector<int>& arr) {
    if (arr.empty()) {
        throw invalid_argument("Array cannot be empty");
    }
    
    int minVal = arr[0];
    for (int i = 1; i < (int)arr.size(); ++i) {
        if (arr[i] < minVal) {
            minVal = arr[i];
        }
    }
    
    return minVal;
}

int main() {
    // Test cases
    cout << findMinimum({3, 1, 4, 1, 5}) << endl; // 1
    cout << findMinimum({10, 20, 5, 30, 2}) << endl; // 2
    cout << findMinimum({7}) << endl; // 7
    cout << findMinimum({5, 5, 5}) << endl; // 5
    cout << findMinimum({-10, 0, 5, -3}) << endl; // -10
    
    return 0;
}
